### PR TITLE
Federation: Qualify Conversation IDs

### PR DESCRIPTION
**Pull request by mdimjasevic**
This patch makes a conversation ID qualified. The now makes it impossible to parse from JSON with the old format because a new field `qualified_id` is expected. Conversely, serialising to JSON retains the old `id` field, but in addition also includes `qualified_id`.

* The gist of the PR is in [Conversation.hs](https://github.com/wireapp/wire-server/pull/1640/files#diff-a3da6271c27cfbacc7d2bb7604cbe22f4a8cb336e9bb4a0807b443ac26290e92); the rest of the PR follows mechanically from it.
* Lots of golden tests had to be updated
* Some integration tests had to be updated

## Checklist

 - [x] Title of this PR explains the impact of the change.
 - [x] The description provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] The CHANGELOG.md file in the *Unreleased* section has been updated to explain the change which will be included in the release notes.
 - [x] If a component uses a new or changed internal endpoint of another component, this is mentioned in the CHANGELOG.md.
- [x] If this PR creates a new endpoint, or adds a new configuration flag, the endpoint / config-flag checklist (see Wire-employee only backend [wiki page](https://github.com/zinfra/backend-wiki/wiki/Checklists)) has been followed.

### CODE REVIEW

**Review by pcapriotti — APPROVED** [review]
I haven't looked at all the changes too carefully, but it looks good. One thing to be generally aware of is that this in an incompatible change at the level of the federation API. Not a problem at this stage, I think, but it might be worth pointing this out in the CHANGELOG.

File: services/galley/test/integration/API.hs [hunk 1]
```diff
@@ -1101,8 +1102,11 @@ listConvsPagingOk = do
...
   walk carl [3, 3, 3, 3, 1] -- 11 (group) + 1 (1:1) + 1 (self)
   where
+    walk :: Foldable t => UserId -> t Int -> TestM ()
     walk u = foldM_ (next u 3) Nothing
+    next :: UserId -> Int32 -> Maybe ConvId -> Int -> TestM (Maybe ConvId)
     next u step start n = do
+      -- TODO(md): Check if the FUTUREWORK note is addressed.
       -- FUTUREWORK: support an endpoint to get qualified conversation IDs
       -- (without all the conversation metadata)
       r1 <- getConvIds u (Right <$> start) (Just step) <!! const 200 === statusCode
```
Comment (mdimjasevic) [inline services/galley/test/integration/API.hs:1112]: @jschaul , I am not sure what you had in mind here. Can you please clarify?

Comment (jschaul) [reply services/galley/test/integration/API.hs:1112]: this futurework comment is unrelated to this PR. It is about creating a new endpoint similar to `GET ​/conversations​/ids Get all conversation IDs.` ([swagger](https://staging-nginz-https.zinfra.io/api/swagger-ui/#/default/get_conversations_ids)).

File: CHANGELOG.md [hunk 2]
```diff
@@ -18,6 +18,8 @@
 
 ## API Changes
 
+* All endpoints that fetch conversation details now also include a new key
+  `qualified_id` for a qualified convevrsation ID (#1640)
```
Comment (jschaul) [inline CHANGELOG.md:22]:
```suggestion
  `qualified_id` for a qualified conversation ID (#1640)
```

**Review by jschaul — COMMENTED** [review]
The golden tests could add this counter-example:

- parsing a json object without a qualified id fails (probably could be added in the https://github.com/wireapp/wire-server/blob/develop/libs/wire-api/test/unit/Test/Wire/API/Golden/FromJSON.hs test)

Is it correct that the conversation data type is always only written, an never parsed by galley/brig/etc ? If it's ever parsed in intra calls, then there might be a deployment order.

Otherwise change looks good.

**Comment by mdimjasevic** [comment]
> The golden tests could add this counter-example:
> 
>     * parsing a json object without a qualified id fails (probably could be added in the https://github.com/wireapp/wire-server/blob/develop/libs/wire-api/test/unit/Test/Wire/API/Golden/FromJSON.hs test)
 
Thank you for the suggestion! I just added this test.

**Comment by mdimjasevic** [comment]
> Is it correct that the conversation data type is always only written, an never parsed by galley/brig/etc ? If it's ever parsed in intra calls, then there might be a deployment order.

@jschaul , I scanned the code base and couldn't find a site where it is parsed from JSON.

**Review by jschaul — APPROVED** [review]
Looks good (once CI agrees)!

**Pull request merged**